src/beachbot/manipulators/drive.py

```diff
@@ -81,51 +103,71 @@ def run(self):
                     self.set_target(0, 0)
 
 
-            # estimate difference between current (last) motor command and targets (requested via set_target)
-            # dir menas how fast robot should rotate around own axis
-            # vel means how fast the robot should move forward
-            # See "set_targets" for more info
-            dir_delta = self._target_angular_vel - self._current_angular_vel
-            vel_delta = self._target_velocity - self._current_velocity
+
+            # Estimate motor velocity based on angluar and linear velocity (unbounded):
+            _target_motor_left_speed = self._target_velocity + 0.5 * self._target_angular_vel
+            _target_motor_right_speed = self._target_velocity - 0.5 * self._target_angular_vel
+            
+
+            # def f1(l,a):
+            #     m1 = l + 0.5*a
+            #     m2 = l - 0.5*a
+            #     desired_power = min(sqrt(a**2 + l**2),100)
+            #     unscaled_max_power = max(abs(m1), abs(m2))
+            #     scale_fac = safe_division(desired_power, unscaled_max_power)
+            #     m1 *= scale_fac
+            #     m2 *= scale_fac
+
+
+
+
+
+            # As we do not have velocity control (we control power, pwm duty cycle for now)
+            # we scale desired output such that total motor power is scaled in relation
+            # to euclidean distance of 2d control input (velocity and angular velocity) to its neutral point (0,0) 
+            # So distance of the joystick form center controls max motor power
+            # Range can be 0-100 percent max motor power:
+            #
+            # 1. desired_power based on, e.g. distance of joystick form center:
+            desired_power = min(sqrt(self._target_angular_vel**2 + self._target_velocity**2),100)
+            # 2. scaling to bound motor power to 100, but also to keep relative mottor values through scaling:
+            unscaled_max_power = max(abs(_target_motor_right_speed), abs(_target_motor_left_speed))
+            # 3. apply both scaling factors, also consider case unscaled_max_power=0 (velocity=0 and anglular_velocity=0) by 'save_division'
+            scale_fac = safe_division(desired_power, unscaled_max_power)
+            _target_motor_left_speed *= scale_fac
+            _target_motor_right_speed *= scale_fac
```

Comment:
Why can't we just use the values from the joystick to directly control the output values? It doesn't matter if it is velocity, power, or any other physical metric. We have PWM values from -100 to 100. I want to control the motors with these values directly. 

I'm completely lost as to why all these calculations are necessary, and no longer have any intuition as to how things should work.

Reply:
I'm not sure how your comment relates to the code highlighted here. `desired_power`, `unscaled_max_power`?